loadtest/user/userentity/actions.go

```diff
@@ -11,28 +11,120 @@ import (
+type authOpenIDAction string
+
+const (
+	authOpenIDLogin  authOpenIDAction = "login"
+	authOpenIDSignup authOpenIDAction = "signup"
+)
+
+var (
+	openIDLoginFormActionRegex = regexp.MustCompile(`action=["'](.*?)["']`)
+)
+
 // SignUp signs up the user with the given credentials.
 func (ue *UserEntity) SignUp(email, username, password string) error {
-	user := model.User{
-		Email:    email,
-		Username: username,
-		Password: password,
+	var newUser *model.User
+	var err error
+
+	switch ue.config.AuthenticationType {
+	case AuthenticationTypeOpenID:
+		if err := ue.authOpenID(authOpenIDSignup); err != nil {
+			return fmt.Errorf("error while signing up using OpenID: %w", err)
+		}
+
+		newUser, _, err = ue.client.GetUserByUsername(context.Background(), username, "")
+		if err != nil {
+			return fmt.Errorf("error while getting user by username: %w", err)
+		}
+	default:
+		user := model.User{
+			Email:    email,
+			Username: username,
+			Password: password,
+		}
+
+		newUser, _, err = ue.client.CreateUser(context.Background(), &user)
+		if err != nil {
+			return err
+		}
+
+		newUser.Password = password
 	}
+	return ue.store.SetUser(newUser)
+}
 
-	newUser, _, err := ue.client.CreateUser(context.Background(), &user)
+// authOpenID logs the user in using OpenID.
+func (ue *UserEntity) authOpenID(action authOpenIDAction) error {
+	if action != authOpenIDLogin && action != authOpenIDSignup {
+		return errors.New("invalid openid auth action")
+	}
+
+	jar, err := cookiejar.New(nil)
 	if err != nil {
-		return err
+		return fmt.Errorf("error while creating cookie jar: %w", err)
 	}
 
-	newUser.Password = password
-	return ue.store.SetUser(newUser)
+	client := &http.Client{
+		Jar: jar,
+	}
+
+	// make a request to the openid login page of mattermost
+	resp, err := client.Get(ue.client.URL + "/oauth/openid/" + string(action))
+	if err != nil {
+		return fmt.Errorf("error while making request to openid %s page: %w", action, err)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("error while reading response body: %w", err)
+	}
+	resp.Body.Close()
+
+	loginURL := string(openIDLoginFormActionRegex.FindSubmatch(body)[1])
+
+	loginResponse, err := client.PostForm(loginURL, url.Values{
+		"username": {ue.config.Username},
+		"password": {ue.config.Password},
+	})
+	if err != nil {
+		return fmt.Errorf("error while %s in: %w", action, err)
+	}
+
+	if loginResponse.StatusCode != http.StatusOK {
+		return fmt.Errorf("%s failed with status code %d", action, loginResponse.StatusCode)
+	}
+
+	cookies := jar.Cookies(loginResponse.Request.URL)
+	for _, cookie := range cookies {
+		if cookie.Name == "MMAUTHTOKEN" {
+			ue.client.SetToken(cookie.Value)
+
+			me, _, err := ue.client.GetMe(context.Background(), "")
+			if err != nil {
+				return fmt.Errorf("error while getting user: %w", err)
+			}
+
+			if err := ue.store.SetUser(me); err != nil {
+				return fmt.Errorf("error while setting user: %w", err)
+			}
```

Comment:
Do we need to do this here? I'd expect this method to purely authenticate (e.g. set the token) and then deal with everything else in `SignUp/Login`.

Reply:
There's a mix of logic here, is either I get the user and set it here, or I get the user by username later and set it in the other method. I've set a TODO to change the logic in a separate PR since I don't want to compromise or delay the test we're going to run today.